File: src/canvas_mcp/core/user_root.py
"""Per-user data directory resolution (local-first).

macOS:  ~/Library/Application Support/{ProductName}/{user_id}/
Windows: %APPDATA%\\{ProductName}\\{user_id}\\
Linux:  ~/.local/share/{ProductName}/{user_id}/

``DEV_USER_ROOT`` overrides the entire user root (useful for local development
and tests without writing under the OS app-support path).
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

from .tenants import PRODUCT_NAME

# Subdirectories created under every user root.
USER_SUBDIRS = (
    "inbox",
    "inbox/courses",
    "inbox/captures",
    "calibration",
    "skills/active",
    "skills/provisional",
    "semantic",
    "auth",
    "sensors",
)


def default_app_support_root(product: str = PRODUCT_NAME) -> Path:
    """OS-standard application support directory for the product."""
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Application Support" / product
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA")
        if appdata:
            return Path(appdata) / product
        return Path.home() / "AppData" / "Roaming" / product
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / product
    return Path.home() / ".local" / "share" / product


def resolve_user_root(
    user_id: str,
    *,
    product: str = PRODUCT_NAME,
    create: bool = False,
) -> Path:
    """Resolve the per-user data directory.

    Honors ``DEV_USER_ROOT`` when set (absolute path). Otherwise
    ``{app_support}/{user_id}/``.
    """
    override = os.environ.get("DEV_USER_ROOT", "").strip()
    if override:
        root = Path(override).expanduser().resolve()
    else:
        if not user_id or "/" in user_id or "\\" in user_id or user_id in (".", ".."):
            raise ValueError(f"Invalid user_id: {user_id!r}")
        root = default_app_support_root(product) / user_id

    if create:
        ensure_user_root(root)
    return root


def ensure_user_root(root: Path) -> Path:
    """Create the standard subdirectory tree under ``root``."""
    root.mkdir(parents=True, exist_ok=True)
    for rel in USER_SUBDIRS:
        (root / rel).mkdir(parents=True, exist_ok=True)
    # Touch empty ledger if missing (append-only file).
    ledger = root / "ledger.jsonl"
    if not ledger.exists():
        ledger.touch()
    return root


def user_path(user_id: str, *parts: str, create_root: bool = False) -> Path:
    """Convenience: ``resolve_user_root(user_id) / parts``."""
    root = resolve_user_root(user_id, create=create_root)
    candidate = root.joinpath(*parts)
    # Callers use this helper for user-owned data.  Reject absolute paths and
    # traversal rather than allowing Path.joinpath() to silently discard the
    # trusted root or escape it through ``..``.
    try:
        candidate.resolve().relative_to(root.resolve())
    except ValueError as exc:
        raise ValueError("user path must remain inside the user root") from exc
    return candidate

File: browser/scripts/lib/user-root.mjs
/**
 * Per-user data directory — mirrors Python ``canvas_mcp.core.user_root``
 * and Tauri ``app/src-tauri/src/inbox.rs``.
 *
 * Priority:
 *   1. ``DEV_USER_ROOT`` (absolute override)
 *   2. OS app-support / ProductName / {PRODUCT_USER_ID || "dev"}
 *
 * Never falls back to ``{repo}/inbox`` — that silent split broke the
 * SSO → API → inbox truth path.
 */
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

export const PRODUCT_NAME = process.env.PRODUCT_NAME?.trim() || "ProductName";

export function resolveProductUserId() {
  const fromEnv = process.env.PRODUCT_USER_ID?.trim();
  return fromEnv || "dev";
}

/** OS-standard application support directory for the product (no user id). */
export function defaultAppSupportRoot(product = PRODUCT_NAME) {
  const home = os.homedir();
  if (process.platform === "darwin") {
    return path.join(home, "Library", "Application Support", product);
  }
  if (process.platform === "win32") {
    const appdata = process.env.APPDATA;
    if (appdata) return path.join(appdata, product);
    return path.join(home, "AppData", "Roaming", product);
  }
  const xdg = process.env.XDG_DATA_HOME;
  if (xdg) return path.join(xdg, product);
  return path.join(home, ".local", "share", product);
}

/**
 * Resolve the per-user data directory (same rules as Python resolve_user_root).
 * @param {{ create?: boolean, announce?: boolean }} [opts]
 */
export function resolveUserRoot(opts = {}) {
  const { create = false, announce = false } = opts;
  const override = process.env.DEV_USER_ROOT?.trim();
  let root;
  let via;
  if (override) {
    root = path.resolve(override);
    via = "DEV_USER_ROOT";
  } else {
    const userId = resolveProductUserId();
    root = path.join(defaultAppSupportRoot(), userId);
    via = `app-support/${userId}`;
  }

  if (create) {
    fs.mkdirSync(path.join(root, "inbox", "courses"), { recursive: true });
    fs.mkdirSync(path.join(root, "calibration"), { recursive: true });
  }

  if (announce) {
    console.error(`[user-root] ${root} (via ${via})`);
  }
  return root;
}

export function resolveInboxDir(opts = {}) {
  return path.join(resolveUserRoot(opts), "inbox");
}
